gen-ai/orchestrator-server/src/main/python/server/src/gen_ai_orchestrator/services/langchain/qa_chain.py

```diff
@@ -0,0 +1,120 @@
+#   limitations under the License.
+#
+"""
+Module for the QA Chain
+It uses LangChain to perform a Conversational Retrieval Chain
+"""
+
+import logging
+import time
+from operator import itemgetter
+
+from langchain_core.retrievers import BaseRetriever
+from langchain_core.runnables import Runnable
+
+from gen_ai_orchestrator.errors.handlers.openai.openai_exception_handler import (
+    openai_exception_handler,
+)
+from gen_ai_orchestrator.errors.handlers.opensearch.opensearch_exception_handler import (
+    opensearch_exception_handler,
+)
+from gen_ai_orchestrator.models.rag.rag_models import Source
+from gen_ai_orchestrator.models.vector_stores.vectore_store_provider import (
+    VectorStoreProvider,
+)
+from gen_ai_orchestrator.routers.requests.requests import QAQuery
+from gen_ai_orchestrator.routers.responses.responses import QAResponse
+from gen_ai_orchestrator.services.langchain.factories.langchain_factory import (
+    get_em_factory,
+    get_vector_store_factory,
+)
+
+logger = logging.getLogger(__name__)
+
+
+@opensearch_exception_handler
+@openai_exception_handler(provider='OpenAI or AzureOpenAIService')
+async def execute_qa_chain(query: QAQuery) -> QAResponse:
+    """
+    Execute the QA chain using the specified embedding settings in the query.
+
+    Args:
+        query (QAQuery): The QA query containing the necessary information for question-answering.
+
+    Returns:
+        QAResponse: The QA response with the document sources.
+
+    """
+    logger.info('QA chain - Start of execution...')
+    start_time = time.time()
+
+    conversational_qa_chain = create_qa_chain(query)
+
+    response = await conversational_qa_chain.ainvoke(
+        **query.question_answering_prompt_inputs
+    )
+
+    qa_duration = '{:.2f}'.format(time.time() - start_time)
+    logger.info('QA chain - End of execution. (Duration : %s seconds)', qa_duration)
+
+    return QAResponse(
+        documents=set(
+            map(
+                lambda doc: Source(
+                    title=doc.metadata['title'],
+                    url=doc.metadata['url'],
+                    content=doc.page_content,
+                ),
+                response,
+            )
+        ),
+    )
+
+
+def build_chain(retriever: BaseRetriever) -> Runnable:
+    """
+    Create the QA chain using the specified retriever.
+
+    Args:
+        retriever: The retriever used in the QA system
+    Returns:
+        The QA chain.
+    """
+    return itemgetter('question') | retriever
+
+
+def create_qa_chain(query: QAQuery) -> Runnable:
+    """
+    Create the QA chain from QAQuery, using the Embedding settings specified in the query
+
+    Args:
+        query: The QA query
+    Returns:
+        The QA chain.
+    """
+    em_factory = get_em_factory(setting=query.embedding_question_em_setting)
+    vector_store = get_vector_store_factory(
+        vector_store_provider=VectorStoreProvider.OPEN_SEARCH,
```

Comment:
Ok for it but I'm reporting on this PR that we should take care of the vector store factory instanciation here in the refactoring of the vector store DB settings. 

TODO AFTER REVIEW SUBMIT @Benvii

Reply:
Comment added on the PR that will be impact by this change.